[TEAM1] Data logging (#37)
SOMAS2021/SOMAS-Final-Report: 003_data_logging/003_data_logging.tex

\chapter{Data Logging and Dashboard}\label{data_logging}

> \section{Why do we need data?}

Reviewer: ```suggestion
\section{Need for Data}
```

Probably best to keep it formal

> \section{Why do we need data?}

Reviewer 2: ```suggestion
\section{Why Do We Need Data?}
```

> Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 

Reviewer 2: ```suggestion
Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messaging frequency, food variation over time, agent death rates, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 
```

> Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc... Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 

Reviewer 3: ```suggestion
Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc.. Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 
```

> Data logging enables analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question: messages frequency, food variation over time, agent deaths rate, etc... Analysis also requires linking different simulation parameters with the data logged (e.g. the number of agents or the amount of food available). 

Reviewer: ```suggestion
Data logging allows for the analysis of how internal variables change with time during a simulation. One might be interested in different variables from the system in question, such as messaging frequency, food variation over time or the agents' death rate, with analysis also requiring the linking of different simulation parameters with their logged data (e.g., the number of agents or the amount of food available). 
```

I think this flows better

> It is therefore important to setup a logging framework that structures data effectively and scales efficiently for large simulations. The below subsections will introduce the simulation parametrisation, followed by the different logging strategies (dumps and structured), and concluding with the logging interface and the analysis dashboard.

Reviewer 2: ```suggestion
It is therefore important to setup a logging framework that structures data effectively and scales efficiently for large simulations. The below subsections will introduce the simulation parameterisation, followed by the different logging strategies (dumps and structured), and concluding with the logging interface and the analysis dashboard.
```

> \section{Parametrisation}

Reviewer 2: ```suggestion
\section{Parameterisation}
```

> First, it was important to enable an effective parametrisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment and a configuration object would subsequently be passed to the simulation environment instance to define the following simulation parameters:

Reviewer 2: ```suggestion
First, it was important to enable an effective parameterisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment and a configuration object is passed to the simulation environment instance to define the following simulation parameters:
```

> First, it was important to enable an effective parametrisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment. A configuration object would therefore be passed to the simulation environment instance, and it would define the following simulation parameters:

Reviewer 2: ```suggestion
First, it was important to enable an effective parameterisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment. A configuration object is passed to the simulation environment instance, and defines the following simulation parameters:
```

> First, it was important to enable an effective parametrisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment. A configuration object would therefore be passed to the simulation environment instance, and it would define the following simulation parameters:

Reviewer: ```suggestion
First, it was important to enable an effective parametrisation of the simulation. A configuration interface was therefore set to abstract variables from the simulation environment and a configuration object would subsequently be passed to the simulation environment instance to define the following simulation parameters:
```

Again, let me know if this is better. Prevents the staggered sentences and repetition of 'therefore'

\begin{multicols}{2}
    \begin{enumerate}
        \item Agent count for each team
        \item Agent count for random agents
        \item Agent count for selfish agents
        \item Agent HP
        \item Agents per floor
        \item Amount of food on the platform
        \item Food per agent ratio (optional)
        \item Ticks per floor
        \item Total days
        \item Reshuffle period
    \end{enumerate}
\end{multicols}

> The configuration parameters could then be written in a simple \texttt{json} file which would be parsed into a configuration object.

Reviewer 2: ```suggestion
The configuration parameters are then written in a JSON file which can be parsed into a configuration object.
```

> \section{General log dumps}

Reviewer 2: ```suggestion
\section{General Log Dumps}
```

> Following the configuration work, a general logger was setup. The general logger is basically a single shared logger between the different instances in the system (simulation, tower, agents, etc...). Each instance appends by default an identifier to a log line in the general logger log file (named \texttt{main.json}). Here is an example below:

Reviewer: ```suggestion
Following the configuration work, a general logger was set up. The general logger is effectively a single shared logger between the different instances in the system (simulation, tower, agents, etc...). Each instance appends by default an identifier to a log line in the general logger log file (named \texttt{main.json}). Here is an example below:
```

\begin{verbatim}
{"agent_id":"dece1252","agent_type":"Team5","floor":3,"health":100,
"level":"info","msg":"Reporting agent state of team 5 agent","reporter":"agent"}

{"Floor:":7,"HP:":100,"Social motive:":"Altruist","agent_id":"2bfd3cfb",
"agent_type":"Team6","level":"info","msg":"Reporting agent state:","reporter":"agent"}

{"agent_id":"7873bb50","agent_type":"Team1","floor":5,"health":100,
"level":"info","msg":"Reporting agent state","reporter":"agent"}
\end{verbatim}

> The general log dumps are unstructured - yet they enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.

Reviewer 2: ```suggestion
The general log dumps are unstructured, but enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.
```

> The general log dumps are not structured yet they enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.

Reviewer 3: ```suggestion
The general log dumps are unstructured - yet they enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.
```

> The general log dumps are not structured yet they enable agent teams to directly log custom states from their agents and use those in e.g. evolutionary algorithms or machine learning training.

Reviewer: ```suggestion
The general log dumps are not structured, yet they enable agent teams to directly log custom states from their agents and use them, for example, in evolutionary algorithms or machine learning training.
```

> \section{Structured logs}

Reviewer 2: ```suggestion
\section{Structured Logs}
```

> The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. After discussions with the agent teams and relating as fit to the problem, the following structured logs where included:

Reviewer 2: ```suggestion
The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data are captured optimally to generate lightweight, understandable logs. After discussions with agent teams, the following structured logs where included:
```

> The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. Following consultations with the agent teams and relating as fit to the problem, the following structured logs where included:

Reviewer 3: ```suggestion
The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. After discussions with the agent teams and relating as fit to the problem, the following structured logs where included:
```

> The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. Following consultations with the agent teams and relating as fit to the problem, the following structured logs where included:

Reviewer: ```suggestion
The general log generates huge amounts of data and therefore was ineffective for data analysis and drawing conclusions easily from different simulations. Therefore, structured logs had to be introduced, where particular data is captured optimally to generate lightweight, understandable logs. Following consultations with the agent teams and relating as fit to the problem, the following structured logs were included:
```

\begin{enumerate}
    \item \textbf{Food}: food available on the platform and food taken by agents.
    \item \textbf{Death}: agent death (including agent state at death time)
    \item \textbf{Messages}: receiver, sender, message type and content
    \item \textbf{Utility}: agent utility (including agent state)
    \item \textbf{Story}: a combination of the above logs in a sequential manner
\end{enumerate}

> The agent state encapsulates by default the agent age, floor, HP, utility, and custom state (e.g. emotions or memory).

Reviewer: ```suggestion
The agent state encapsulates by default the agent's age, floor, HP, utility, and custom state (e.g., emotions or memory).
```

Author: Do we need the comma in `(e.g., emotions or memory).` or typo?

Reviewer: we do - agreed on slack last night

> In the general log, custom agents could pass all sorts of information and therefore the log schema could not determined at compile time (which would be required for a visualisation dashboard). Structured logs have a well defined schema which does not depend on custom agents. For example, a food log will always take the following form:

Reviewer 2: ```suggestion
In the general log, custom agents could pass all sorts of information and therefore the log schema could not determined at compile time (which is required for a visualisation dashboard). Structured logs have a well-defined schema which does not depend on custom agents. For example, a food log will always take the following form:
```

> In the general log, custom agents could pass all sorts of information and therefore the log schema could not determined at compile time (which would be required for a visualisation dashboard). Structured logs have a well defined schema which does not depend on custom agents. For example, a food log will always have this form:

Reviewer 3: ```suggestion
In the general log, custom agents could pass all sorts of information and therefore the log schema could not determined at compile time (which would be required for a visualisation dashboard). Structured logs have a well defined schema which does not depend on custom agents. For example, a food log will always take the following form:
```

\begin{verbatim}
{"day":2,"food":19,"level":"info","tick":222,"time":"2022-01-17T02:13:37Z"}
{"day":2,"food":0,"level":"info","tick":262,"time":"2022-01-17T02:13:37Z"}
{"day":3,"food":100,"level":"info","tick":321,"time":"2022-01-17T02:13:37Z"}
{"day":3,"food":90,"level":"info","tick":322,"time":"2022-01-17T02:13:37Z"}
\end{verbatim}

> With structured logs setup, it is now possible to bundle logging into an interface and connect it to a visualisation tool.

Reviewer 3: ```suggestion
With structured logs set up, it is now possible to bundle logging into an interface and connect it to a visualisation tool.
```

> \section{Logging interface}

Reviewer 2: ```suggestion
\section{Logging Interface}
```

> It was important to set up logging such as structured loggers could be easily extended with new loggers, while preserving the ability to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.

Reviewer 2: ```suggestion
It was important to set up logging such that structured loggers could be easily extended with new loggers, while preserving the ability to generate general logs. The following design was adopted to encapsulate the simulation configuration and the loggers.
```

> It was important to setup logging such as structured loggers could be easily extended with new loggers, while preserving the possibility to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.

Reviewer 3: ```suggestion
It was important to set up logging such as structured loggers could be easily extended with new loggers, while preserving the ability to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.
```

> It was important to setup logging such as structured loggers could be easily extended with new loggers, while preserving the possibility to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.

Reviewer: ```suggestion
It was important to set up logging such that structured loggers could be easily extended with new loggers, while preserving the possibility to generate general logs. The following design has been adopted to encapsulate the simulation configuration and the loggers.
```

\begin{figure}[htb]
    \centering
    \includegraphics[width=0.8\linewidth]{003_data_logging/images/struct.png}
    \caption{Diagram showing how logging fits in the system}
    \label{fig:design_logging}
\end{figure}

> The entry point to the simulation dispatcher was then linked to both the CLI as well as a server (with HTTP listeners) to support a browser-based dashboard.

Reviewer 2: ```suggestion
The entry point to the simulation dispatcher is linked to both the command line interface as well as a server (with HTTP listeners) to support a browser-based dashboard.
```
